### PR TITLE
New segmentation model: GM and WM for exvivo DWI data (University of Queensland)

### LINKED ISSUE #3004: Generate a deepseg model for exvivo spinal cord
Context: https://forum.spinalcordmri.org/t/error-when-running-sct-get-centerline/551/8

This user privately shared beautiful DWI data, see an example here:
Screen Shot 2020-11-04 at 9 55 46 PM

We need to generate a specific deepseg model to segment the spinal cord and gray matter. I believe @charleygros has done most of the work for his previous exvivo spinalcord segmentation (ISMRM submission).

### PR DESCRIPTION
**Pull request by charleygros**
This PR integrates a new segmentation model to `sct_deepseg`.

This model has been trained on MRI exvivo DWI data provided by @ihattan2020 (University of Queensland - Center of Advanced Imaging). Model was trained with `ivadomed`, including the following features: 2D unet, multiclass (grey and white matters), softseg.

Fixes #3004.

### CODE REVIEW

**Comment by charleygros** [comment]
Associated model repository: https://github.com/ivadomed/model_seg_mice_gm-wm_dwi_unet2d-multichannel-softseg

**Review by andreanne-lemay — APPROVED** [review]
The description overflows in the other column. I changed the width to make the description fit here: 189a722
BEFORE:
image
AFTER:
image

Otherwise looks great!

File: spinalcordtoolbox/deepseg/models.py [hunk 1]
```diff
@@ -68,6 +68,14 @@
         "description": "Multiclass cord tumor segmentation model.",
         "contrasts": ["t2", "t1"],
         "default": False,
+    },
+    "model_seg_mice_gm-wm_dwi_unet2d-multichannel-softseg": {
+        "url": [
+            "https://github.com/ivadomed/model_seg_mice_gm-wm_dwi_unet2d-multichannel-softseg/archive/r20210218.zip"
+        ],
+        "description": "Grey and white matters segmentation on exvivo DWI mouse data.",
```
Comment (jcohenadad) [inline spinalcordtoolbox/deepseg/models.py:76]:
```suggestion
        "description": "Grey/white matter seg on exvivo mouse DWI",
```

**Comment by jcohenadad** [comment]
> The description overflows in the other column. I changed the width to make the description fit here: 189a722

the recommended terminal width is 100, and currently this is even too small for deepseg description:
Screen Shot 2021-02-18 at 2 54 09 PM

so instead of further extending columns (we should aim at the opposite), i would suggest to squeeze the description https://github.com/neuropoly/spinalcordtoolbox/pull/3224#discussion_r578705217

**Comment by charleygros** [comment]
When trying the model I get this error:
```
Traceback (most recent call last):
  File "/home/charley/sct-dev/spinalcordtoolbox/scripts/sct_deepseg.py", line 245, in <module>
    main(sys.argv[1:])
  File "/home/charley/sct-dev/spinalcordtoolbox/scripts/sct_deepseg.py", line 207, in main
    nii_lst, target_lst = imed_inference.segment_volume(path_model, input_filenames, options=options)
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/ivadomed/inference.py", line 255, in segment_volume
    preds = model(img) if fname_model.endswith('.pt') else onnx_inference(fname_model, img)
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/ivadomed/inference.py", line 368, in onnx_inference
    ort_session = onnxruntime.InferenceSession(model_path)
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/onnxruntime/capi/session.py", line 158, in __init__
    self._load_model(providers or [])
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/onnxruntime/capi/session.py", line 177, in _load_model
    self._sess.load_model(providers)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Type Error: Type 'tensor(bool)' of input parameter (281) of operator (ScatterND) in node (ScatterND_160) is invalid.
Traceback (most recent call last):
  File "/home/charley/sct-dev/spinalcordtoolbox/scripts/sct_deepseg.py", line 245, in <module>
    main(sys.argv[1:])
  File "/home/charley/sct-dev/spinalcordtoolbox/scripts/sct_deepseg.py", line 207, in main
    nii_lst, target_lst = imed_inference.segment_volume(path_model, input_filenames, options=options)
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/ivadomed/inference.py", line 255, in segment_volume
    preds = model(img) if fname_model.endswith('.pt') else onnx_inference(fname_model, img)
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/ivadomed/inference.py", line 368, in onnx_inference
    ort_session = onnxruntime.InferenceSession(model_path)
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/onnxruntime/capi/session.py", line 158, in __init__
    self._load_model(providers or [])
  File "/home/charley/sct-dev/python/envs/venv_sct/lib/python3.6/site-packages/onnxruntime/capi/session.py", line 177, in _load_model
    self._sess.load_model(providers)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Type Error: Type 'tensor(bool)' of input parameter (281) of operator (ScatterND) in node (ScatterND_160) is invalid.
```

**Comment by charleygros** [comment]
Trying to replace the onnx model by its pt version.

**Comment by charleygros** [comment]
Update: I am restarting the training. Sorry about that.

**Comment by charleygros** [comment]
> Update: I am restarting the training. Sorry about that.

AH! The problem was actually coming from a bug in `ivadomed`, see [PR]([url](https://github.com/ivadomed/ivadomed/pull/688)), which is, I think, fixed.

Note: we may want to do a new ivadomed release before the SCT one in order to incorporate these changes :-)

**Comment by charleygros** [comment]
Please do not merge this PR yet: still working on the model (release 20210220) --> once properly tested and uploaded on its repo, it will be good to go! Keep posted soon

**Comment by charleygros** [comment]
Release [done](https://github.com/ivadomed/model_seg_mice_gm-wm_dwi_unet2d-multichannel-softseg/releases/tag/r20210220)!

**Pull request merged**